src/test/java/org/elasticsearch/cluster/ClusterInfoServiceTests.java

```diff
@@ -0,0 +1,259 @@
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.plugins.AbstractPlugin;
+import org.elasticsearch.test.ElasticsearchIntegrationTest;
+import org.elasticsearch.test.InternalTestCluster;
+import org.elasticsearch.test.transport.MockTransportService;
+import org.elasticsearch.transport.*;
+import org.hamcrest.Matchers;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.greaterThanOrEqualTo;
+
+/**
+ * Integration tests for the ClusterInfoService collecting information
+ */
+@ElasticsearchIntegrationTest.ClusterScope(scope = ElasticsearchIntegrationTest.Scope.TEST, numDataNodes = 0)
+public class ClusterInfoServiceTests extends ElasticsearchIntegrationTest {
+
+    public static class Plugin extends AbstractPlugin {
+
+        @Override
+        public String name() {
+            return "ClusterInfoServiceTests";
+        }
+
+        @Override
+        public String description() {
+            return "ClusterInfoServiceTests";
+        }
+
+        public void onModule(ActionModule module) {
+            module.registerFilter(BlockingActionFilter.class);
+        }
+    }
+
+    public static class BlockingActionFilter extends org.elasticsearch.action.support.ActionFilter.Simple {
+
+        ImmutableSet<String> blockedActions = ImmutableSet.of();
+
+        @Inject
+        public BlockingActionFilter(Settings settings) {
+            super(settings);
+        }
+
+        @Override
+        protected boolean apply(String action, ActionRequest request, ActionListener listener) {
+            if (blockedActions.contains(action)) {
+                throw new ElasticsearchException("force exception on [" + action + "]");
+            }
+            return true;
+        }
+
+        @Override
+        protected boolean apply(String action, ActionResponse response, ActionListener listener) {
+            return true;
+        }
+
+        @Override
+        public int order() {
+            return 0;
+        }
+
+        public void blockActions(String... actions) {
+            HashSet<String> newSet = new HashSet<>(blockedActions);
+            newSet.addAll(Arrays.asList(actions));
+            blockedActions = ImmutableSet.copyOf(newSet);
+        }
+    }
+
+    static class InfoListener implements ClusterInfoService.Listener {
+        final AtomicReference<CountDownLatch> collected = new AtomicReference<>(new CountDownLatch(1));
+        final AtomicReference<ClusterInfo> refInfo = new AtomicReference<>();
```

Comment:
It doesn't look like you are using any of the regular `getThenSet` features of the AtomicReference, can this just be a mutable variable? (Not really a big deal either way)